third_party/move/move-compiler-v2/src/pipeline/lint_processor/needless_mutable_reference.rs

```diff
@@ -0,0 +1,236 @@
+                debug_assert!(srcs.len() == 1 && dsts.len() == 1);
+                if self.node_exists(srcs[0])
+                    && target.get_local_type(dsts[0]).is_mutable_reference()
+                {
+                    self.add_derived_edge(dsts[0], srcs[0]);
+                }
+            },
+            _ => {},
+        }
+    }
+
+    /// Update mutable usages given `instr`.
+    fn update_mutably_used(&mut self, env: &GlobalEnv, instr: &Bytecode) {
+        use Bytecode::*;
+        use Operation::*;
+        match instr {
+            Call(_, _, WriteRef, srcs, _) => {
+                self.set_and_propagate_mutably_used(srcs[0]);
+            },
+            Call(_, _, Function(mid, fid, _), srcs, _) => {
+                let callee_env = env.get_function_qid(mid.qualified(*fid));
+                callee_env
+                    .get_parameter_types()
+                    .iter()
+                    .enumerate()
+                    .filter(|(_, ty)| ty.is_mutable_reference())
+                    .map(|(i, _)| srcs[i])
+                    .for_each(|src| self.set_and_propagate_mutably_used(src));
+            },
+            _ => {},
+        }
+    }
+
+    /// Add an origin to the tracker.
+    fn add_origin(&mut self, origin: TempIndex, loc: Loc) {
+        self.origins.entry(origin).or_insert(loc);
+        self.derived_edges.entry(origin).or_default();
+    }
+
+    /// Check if a node exists in the tracker.
+    fn node_exists(&self, node: TempIndex) -> bool {
+        self.derived_edges.contains_key(&node)
+    }
+
+    /// Add a derived edge to the tracker.
+    fn add_derived_edge(&mut self, from: TempIndex, to: TempIndex) {
+        self.derived_edges.entry(from).or_default().insert(to);
+        self.propagate_mutably_used(from, to);
+    }
+
+    /// Propagate mutably used information from `from` to `to`.
+    fn propagate_mutably_used(&mut self, from: TempIndex, to: TempIndex) {
+        if self.mutably_used.contains(&from) {
+            self.set_and_propagate_mutably_used(to);
+        }
+    }
+
+    /// Set a mutable reference as mutably used.
+    /// Propagate this information transitively through derived edges.
+    /// Propagation is stopped early if a node is already marked as mutably used.
+    fn set_and_propagate_mutably_used(&mut self, node: TempIndex) {
+        let mut mutably_used = std::mem::take(&mut self.mutably_used);
+        self._set_and_propagate_mutably_used(node, &mut mutably_used);
+        self.mutably_used = mutably_used;
+    }
+
+    /// Helper function for `set_and_propagate_mutably_used`.
+    fn _set_and_propagate_mutably_used(
+        &self,
+        node: TempIndex,
+        mutably_used: &mut BTreeSet<TempIndex>,
+    ) {
+        if !mutably_used.insert(node) {
+            // Stop early if a node is already marked as mutably used.
+            return;
+        }
+        if let Some(parents) = self.derived_edges.get(&node) {
```

Comment:
nit: could you add a comment saying that the graph can never be cyclic?

Reply:
Is that relevant (here)? Note that we have the stop early check above, which would prevent an infinite recursion cycle.